dhall/src/Dhall/Eval.hs

```diff
@@ -0,0 +1,900 @@
+
+inst :: Eq a => Closure a -> Val a -> Val a
+inst (Cl x env t) !u = eval (Extend env x u) t
+{-# inline inst #-}
+
+-- Out-of-env variables have negative de Bruijn levels.
+vVar :: Env a -> Var -> Val a
+vVar env (V x (fromInteger -> i :: Int)) = go env i where
+  go (Extend env x' v) i
+    | x == x'   = if i == 0 then v else go env (i - 1)
+    | otherwise = go env i
+  go (Skip env x') i
+    | x == x'   = if i == 0 then VVar x (countName x env) else go env (i - 1)
+    | otherwise = go env i
+  go Empty i = VVar x (0 - i - 1)
+
+vApp :: Eq a => Val a -> Val a -> Val a
+vApp !t !u = case t of
+  VLam _ t    -> inst t u
+  VHLam _ t   -> t u
+  t           -> VApp t u
+{-# inline vApp #-}
+
+vCombine :: Val a -> Val a -> Val a
+vCombine t u = case (t, u) of
+  (VRecordLit m, u) | null m    -> u
+  (t, VRecordLit m) | null m    -> t
+  (VRecordLit m, VRecordLit m') -> VRecordLit (Dhall.Map.sort (Dhall.Map.unionWith vCombine m m'))
+  (t, u)                        -> VCombine t u
+{-# inline vCombine #-}
+
+vCombineTypes :: Val a -> Val a -> Val a
+vCombineTypes t u = case (t, u) of
+  (VRecord m, u) | null m -> u
+  (t, VRecord m) | null m -> t
+  (VRecord m, VRecord m') -> VRecord (Dhall.Map.sort (Dhall.Map.unionWith vCombineTypes m m'))
+  (t, u)                  -> VCombineTypes t u
+{-# inline vCombineTypes #-}
+
+vListAppend :: Val a -> Val a -> Val a
+vListAppend t u = case (t, u) of
+  (VListLit _ xs, u) | null xs   -> u
+  (t, VListLit _ ys) | null ys   -> t
+  (VListLit t xs, VListLit _ ys) -> VListLit t (xs <> ys)
+  (t, u)                         -> VListAppend t u
+{-# inline vListAppend #-}
+
+vNaturalPlus :: Val a -> Val a -> Val a
+vNaturalPlus t u = case (t, u) of
+  (VNaturalLit 0, u            ) -> u
+  (t,             VNaturalLit 0) -> t
+  (VNaturalLit m, VNaturalLit n) -> VNaturalLit (m + n)
+  (t,             u            ) -> VNaturalPlus t u
+{-# inline vNaturalPlus #-}
+
+eval :: forall a. Eq a => Env a -> Expr Void a -> Val a
+eval !env t =
+  let
+    evalE :: Expr Void a -> Val a
+    evalE = eval env
+    {-# inline evalE #-}
+
+    evalChunks :: Chunks Void a -> VChunks a
+    evalChunks (Chunks xys z) =
+      foldr' (\(x, t) vcs ->
+                case evalE t of
+                  VTextLit vcs' -> VChunks [] x <> vcs' <> vcs
+                  t             -> VChunks [(x, t)] mempty <> vcs)
+            (VChunks [] z)
+            xys
+    {-# inline evalChunks #-}
+
+  in case t of
+    Const k          -> VConst k
+    Var v            -> vVar env v
+    Lam x a t        -> VLam (evalE a) (Cl x env t)
+    Pi x a b         -> VPi (evalE a) (Cl x env b)
+    App t u          -> vApp (evalE t) (evalE u)
+    Let (b :| bs) t  -> go env (b:bs) where
+                          go !env []     = eval env t
+                          go  env (b:bs) = go (Extend env (variable b)
+                                                          (eval env (value b))) bs
```

Comment:
No, we add new entries to the environment as we process the `Let` block, because definitions in the block and the `Let` body may depend on previous definitions. `evalE` is shorthand only for non-`Env` changing `eval`.